dist/expected.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2016 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package dist
+
+import "github.com/gonum/quad"
+
+// ExpectedFixed numerically computes the expected value of the input function
+// under the probability distribution specified by the Quantiler. That
+// is, ExpectedFixed computes
+//
+//  ev ≈ E[f] = int_-inf^inf f(x) p(x) dx
+//
+// using the number of quadrature locations specified by evals. Concurrent
+// sets the number of allowed concurrent evaluations of f.
+//
+// ExpectedFixed is a wrapper around quad.Fixed. See the documentation for more
+// information.
+func ExpectedFixed(f func(float64) float64, q Quantiler, evals, concurrent int) (ev float64) {
```

Comment:
Your different addition schemes I was classifying as rounding errors, since they give the same answer in infinite precision arithmetic. While I sympathize with your point, I think you're being overly picky. There are lots of cases in gonum where we change minor things -- parentheses around additions, changing native.Daxpy to use SIMD, etc. For "normal" inputs those changes don't matter, they change by a few ULPs. There are definitely cases where there are inputs like you mention, but those inputs need to be using specialized algorithms. In the particular case of quadrature, if the function values at your four quadrature points are  [1e30, 1e30, -2e30, 2.0], the estimate of the integral is completely useless no matter if 0 or 0.5 is returned. More generally, I don't see how we can possibly update code if we have to guarantee an exact double precision match for all inputs for every change.

If this kind of thing is actually a real problem for quadrature routines, we can fix the algorithm so that it does return a ULP identical answer for concurrent and non-concurrent. Right now we use a mutex and add w_i f(x_i), but we could alternately save the w_i and f(x_i) results into a slice and perform a weighted sum once they're all computed.